File: 3/3000s/3507/MinimumPairRemovalToSortArrayI.cpp
/*
https://leetcode.com/problems/minimum-pair-removal-to-sort-array-i/description/?envType=daily-question&envId=2026-01-22
*/

class Solution {
public:

    bool isValid(vector<int>&nums){
        vector<int>tochk=nums;
        sort(tochk.begin(), tochk.end());
        if(tochk==nums) return true;
        return false;
    }

    int minimumPairRemoval(vector<int>& nums) {
        int sz=nums.size(), res=0;
        while(true){
            if(isValid(nums)) break;
            map<int, vector<pair<int, int>>>mp;
            for(int i=0; i<sz; ++i){
                if(i+1<=sz-1){
                    int cursm=nums[i]+nums[i+1];
                    mp[cursm].push_back({i, i+1});;
                }
            }
            auto itr=mp.begin();
            int smToAd=itr->first;
            auto pair=(itr->second)[0];
            int frstIdx=pair.first, secIdx=pair.second;
            vector<int>nwNums;
            int idx=0;
            while(true){
                if(idx>=sz) break;
                if(idx==frstIdx){
                    nwNums.push_back(smToAd);
                    nums[idx]=-1e9;
                    nums[idx+1]=-1e9;
                    idx+=2;
                } else nwNums.push_back(nums[idx++]);
            }
            nums.erase(remove(nums.begin(), nums.end(), -1e9), nums.end());
            nums=nwNums;
            sz=nums.size();
            res++;
        }
        return res;
    }
};

File: 3/3000s/3507/MinimumPairRemovalToSortArrayI.java
/*
https://leetcode.com/problems/minimum-pair-removal-to-sort-array-i/description/?envType=daily-question&envId=2026-01-22
*/

class Pair{

    public int first;

    public int second;

    public Pair(int first, int second){
        this.first=first;
        this.second=second;
    }
}

class Solution {

    public boolean isAlreadySorted(int[] nums){
        int[] numsToCheck=new int[nums.length];
        for(int i=0; i<nums.length; ++i){
            numsToCheck[i]=nums[i];
        }
        Arrays.sort(numsToCheck);
        return Arrays.equals(numsToCheck, nums);
    }

    public int minimumPairRemoval(int[] nums) {
        int lengthOfNums=nums.length, resultantNumberOfOperations=0;
        while(true){
            if(isAlreadySorted(nums)) break;
            TreeMap<Integer, ArrayList<Pair>>sumIndicesMap=new TreeMap<>();
            for(int i=0; i<lengthOfNums; ++i){
                if(i+1<=lengthOfNums-1){
                    int currentPairSum=nums[i]+nums[i+1];
                    if(sumIndicesMap.containsKey(currentPairSum)){
                        ArrayList<Pair>currentPairLst=sumIndicesMap.get(currentPairSum);
                        Pair newPair=new Pair(i, i+1);
                        currentPairLst.add(newPair);
                        sumIndicesMap.put(currentPairSum, currentPairLst);
                    } else{
                        ArrayList<Pair>newPairLst=new ArrayList<>();
                        Pair newPair=new Pair(i, i+1);
                        newPairLst.add(newPair);
                        sumIndicesMap.put(currentPairSum, newPairLst);
                    }
                }
            }
            Integer lowestSum=sumIndicesMap.firstKey();
            ArrayList<Pair>allIndicesForLowestSum=sumIndicesMap.get(lowestSum);
            Pair pairToCheck=allIndicesForLowestSum.get(0);
            int firstIndex=pairToCheck.first, secondIndex=pairToCheck.second;
            int[] newNums=new int[lengthOfNums-1];
            int beginIndex=0, indexForNewNums=0;
            while(true){
                if(beginIndex>=lengthOfNums) break;
                if(beginIndex==firstIndex){
                    newNums[indexForNewNums]=lowestSum;
                    beginIndex+=2;
                } else{
                    newNums[indexForNewNums]=nums[beginIndex++];
                }
                indexForNewNums++;
            }
            nums=newNums;
            lengthOfNums=nums.length;
            resultantNumberOfOperations++;
        }
        return resultantNumberOfOperations;
    }
}